### ISSUE TITLE
date_histogram - context aware extended_bounds

### ISSUE THREAD

**Issue by LucidObscurity** [elastic/elasticsearch#28889]
**Describe the feature**:
```json
  "aggs": {
    "longHisto": {
      "date_histogram": {
        "field": "@timestamp",
        "interval": "1h",
        "time_zone": "America/Los_Angeles",
        "min_doc_count": 0,
        "extended_bounds": {"min": "now/d", "max": "now/d+1d"}
      },
      "aggs": {
        "shortHisto": {
          "date_histogram": {
            "field": "@timestamp",
            "interval": "10m",
            "time_zone": "America/Los_Angeles",
            "min_doc_count": 0,
            "extended_bounds": {"min": "longHisto>key", "max": "longHisto>key+1h"}
          }
        },
        "stats": {
          "extended_stats_bucket": {
              "buckets_path": "shortHisto>_count",
              "sigma":1
          }
        },
        "trueAvg": {
          "bucket_script": {
            "buckets_path": {
              "sum": "stats.sum"
            },
            "script": "params.sum / 6"
          }
        }
      }
    }
```

Ability to reference a parent bucket aggregation to access dynamic date values for use in defining extended_bounds. The "shortHisto" above uses invalid syntax to attempt to represent the use case. 

This would be useful for analyzing the rate of data or the rate of certain types of data over time and would allow extended_stats_bucket to still produce meaningful information even where there are no matching docs for a given sub interval. 

Currently, using extended_stats_bucket this way is only correct if the first and last possible buckets in the "shortHisto" have docs, then "min_doc_count: 0" will fill in the missing buckets, but if you don't have a first and last, you will not get the correct number of buckets for the extended_stats_bucket values to be meaningful.

**Elasticsearch version** (`bin/elasticsearch --version`):
6.2.2

**Comment by LucidObscurity** [comment]
Thanks for taking a look!

Below is a sample of data from the results of the above query. 

To answer your first question, `extended_stats_bucket` is incorrect whenever `shortHisto` has less than 6 buckets (`shortHisto: 10m, longHisto: 1h`). The stats are not entirely incorrect, but any mathematical formula which uses "count" is incorrect, at least in terms of what I'm trying to represent: the average # of documents per 10m for the 1h. What I'm looking to analyze is the rate of documents filtered by some category.

- In the first bucket, only 1 bucket was returned. The `extended_status_bucket` therefore calculates the average as doc_count.sum: 3 / 1 bucket, but there are six 10m windows per 1h window. The average rate of documents per 10m window for the hour should be doc_count.sum: 3 / 6 buckets where 5 buckets have doc_count: 0 (at least for what I'm trying to model). 
- The second and third buckets shows the influence of `min_doc_count: 0` as the "middle" buckets are filled in with `doc_count: 0`. From what I've seen, the stats will be fine so long as the first and last 10m window have at least one document.
- The fourth bucket again shows `min_doc_count: 0` at work but in this case there are no documents for the 40 and 50 minute windows for the hour. This, and the first bucket, is where I wanted to use `extended_bounds` to solve the problem.
- The fifth and last bucket is also a problem as there are no docs at all and thus no buckets (unless I could use `extended_bounds` to force their being filled in.

It's possible this is a bug and not the expected output but being fairly new to working with elasticsearch I though it would be presumptuous to assume any unexpected output was a bug and not just my lack of understanding.

This issue is definitely unique to nesting histograms. I'm a bit surprised it hasn't come up before since it seems to be necessary to model rate of documents over time (as opposed to gathering stats on a value contained in the documents which is trivial and works well).

I agree that any reference to a non-time value in the "buckets_path" type of example I used would be problematic. Do you think this would be more problematic than referencing say a `term` in the `buckets_path` for an aggregation which requires a numeric? Either way, the syntax example I used is not well thought out nor based on any knowledge of the source code and how this type of feature might be best implemented. Maybe a boolean option or something else entirely would be a better approach.

Thanks again.

```json
{
	"key_as_string" : "2018-03-02T03:00:00.000-08:00",
	"key" : 1519988400000,
	"doc_count" : 3,
	"shortHisto" : {
		"buckets" : [{
				"key_as_string" : "2018-03-02T03:00:00.000-08:00",
				"key" : 1519988400000,
				"doc_count" : 3
			}
		]
	},
	"stats" : {
		"count" : 1,
		"min" : 3,
		"max" : 3,
		"avg" : 3,
		"sum" : 3,
		"sum_of_squares" : 9,
		"variance" : 0,
		"std_deviation" : 0,
		"std_deviation_bounds" : {
			"upper" : 3,
			"lower" : 3
		}
	},
	"trueAvg" : {
		"value" : 0.5
	}
}, {
	"key_as_string" : "2018-02-27T16:00:00.000-08:00",
	"key" : 1519776000000,
	"doc_count" : 163,
	"shortHisto" : {
		"buckets" : [{
				"key_as_string" : "2018-02-27T16:00:00.000-08:00",
				"key" : 1519776000000,
				"doc_count" : 3
			}, {
				"key_as_string" : "2018-02-27T16:10:00.000-08:00",
				"key" : 1519776600000,
				"doc_count" : 0
			}, {
				"key_as_string" : "2018-02-27T16:20:00.000-08:00",
				"key" : 1519777200000,
				"doc_count" : 0
			}, {
				"key_as_string" : "2018-02-27T16:30:00.000-08:00",
				"key" : 1519777800000,
				"doc_count" : 0
			}, {
				"key_as_string" : "2018-02-27T16:40:00.000-08:00",
				"key" : 1519778400000,
				"doc_count" : 0
			}, {
				"key_as_string" : "2018-02-27T16:50:00.000-08:00",
				"key" : 1519779000000,
				"doc_count" : 160
			}
		]
	},
	"stats" : {
		"count" : 6,
		"min" : 0,
		"max" : 160,
		"avg" : 27.166666666666668,
		"sum" : 163,
		"sum_of_squares" : 25609,
		"variance" : 3530.1388888888887,
		"std_deviation" : 59.414971925339565,
		"std_deviation_bounds" : {
			"upper" : 86.58163859200623,
			"lower" : -32.2483052586729
		}
	},
	"trueAvg" : {
		"value" : 27.166666666666668
	}
}, {
	"key_as_string" : "2018-03-02T07:00:00.000-08:00",
	"key" : 1520002800000,
	"doc_count" : 897,
	"shortHisto" : {
		"buckets" : [{
				"key_as_string" : "2018-03-02T07:00:00.000-08:00",
				"key" : 1520002800000,
				"doc_count" : 16
			}, {
				"key_as_string" : "2018-03-02T07:10:00.000-08:00",
				"key" : 1520003400000,
				"doc_count" : 0
			}, {
				"key_as_string" : "2018-03-02T07:20:00.000-08:00",
				"key" : 1520004000000,
				"doc_count" : 0
			}, {
				"key_as_string" : "2018-03-02T07:30:00.000-08:00",
				"key" : 1520004600000,
				"doc_count" : 232
			}, {
				"key_as_string" : "2018-03-02T07:40:00.000-08:00",
				"key" : 1520005200000,
				"doc_count" : 325
			}, {
				"key_as_string" : "2018-03-02T07:50:00.000-08:00",
				"key" : 1520005800000,
				"doc_count" : 324
			}
		]
	},
	"stats" : {
		"count" : 6,
		"min" : 0,
		"max" : 325,
		"avg" : 149.5,
		"sum" : 897,
		"sum_of_squares" : 264681,
		"variance" : 21763.25,
		"std_deviation" : 147.5237269051999,
		"std_deviation_bounds" : {
			"upper" : 297.0237269051999,
			"lower" : 1.9762730948000922
		}
	},
	"trueAvg" : {
		"value" : 149.5
	}
}, {
	"key_as_string" : "2018-02-27T22:00:00.000-08:00",
	"key" : 1519797600000,
	"doc_count" : 49,
	"shortHisto" : {
		"buckets" : [{
				"key_as_string" : "2018-02-27T22:00:00.000-08:00",
				"key" : 1519797600000,
				"doc_count" : 3
			}, {
				"key_as_string" : "2018-02-27T22:10:00.000-08:00",
				"key" : 1519798200000,
				"doc_count" : 0
			}, {
				"key_as_string" : "2018-02-27T22:20:00.000-08:00",
				"key" : 1519798800000,
				"doc_count" : 0
			}, {
				"key_as_string" : "2018-02-27T22:30:00.000-08:00",
				"key" : 1519799400000,
				"doc_count" : 46
			}
		]
	},
	"stats" : {
		"count" : 4,
		"min" : 0,
		"max" : 46,
		"avg" : 12.25,
		"sum" : 49,
		"sum_of_squares" : 2125,
		"variance" : 381.1875,
		"std_deviation" : 19.524023663169434,
		"std_deviation_bounds" : {
			"upper" : 31.774023663169434,
			"lower" : -7.274023663169434
		}
	},
	"trueAvg" : {
		"value" : 8.166666666666666
	}
}, {
	"key_as_string" : "2018-03-02T13:00:00.000-08:00",
	"key" : 1520024400000,
	"doc_count" : 0,
	"shortHisto" : {
		"buckets" : []
	},
	"stats" : {
		"count" : 0,
		"min" : null,
		"max" : null,
		"avg" : null,
		"sum" : null,
		"sum_of_squares" : null,
		"variance" : null,
		"std_deviation" : null,
		"std_deviation_bounds" : {
			"upper" : null,
			"lower" : null
		}
	}
}
```

**Comment by polyfractal** [comment]
@LucidObscurity Thanks for the details! (and sorry for the delay, was away for a work trip and traveling home)

I see what you're saying now, makes sense.  Because the second short histo is inside the first, you need to expand the ranges dynamically based on the keys if you need any kind of computation done on those buckets.

Maybe there's a workaround with the current limitations, but I'm not thinking of anything clever at the moment. 

> Do you think this would be more problematic than referencing say a term in the buckets_path for an aggregation which requires a numeric?

Only in the sense that those kind of issues only happen right now in `buckets_path` which is solely found on pipeline aggs.  Whereas this would extend the syntax to another parameter (`extended bounds`) on normal histo/date_histo aggs.  But you're right that it's the same potential error.

I wonder if we could instead do something like `extended_bounds: "parent_bounds"` or something to automate the operation of "extend this internal agg to the width of the parent".  Not sure that is technically feasible though.  We fill empty buckets after the current agg is reduced, but I'm not sure we know the bounds of the parent agg yet.

In any case, I'll bring this issue up at our next search/aggs meeting on monday, I think it's a totally valid enhancement request.  Thanks for the extra details!

**Comment by polyfractal** [comment]
Heya @LucidObscurity.  So we chatted about this, and I'm afraid we don't have a great solution.  We generally agreed this is a valid feature that would be nice to have, but we're conflicted on how to implement it.  No one was overly keen on extending the `extended_bounds` syntax, since that opens up a number of technical issues.  

The main issue is that most aggregations know nothing about the context they live in; they execute their part of the agg tree and call their children aggs to do the same.  There's very little interaction with the parents or children.  Pipeline aggs break this model, and are fairly more complicated (with pointier edge-cases) for that reason.  So that's the main motivation against adding pathing support to `extended_bounds`

Short term, you could probably replace the `extended_stats_bucket` with a `bucket_script` and calculate the values yourself, since you know that there _should_ be six buckets per hourly interval.  Not ideal, and definitely not user-friendly, but it'd get you unstuck for now.

Longer term, we kicked around potentially modifying the pipeline aggs (e.g. `extended_stats_bucket`) to handle this situation, perhaps something like a parameter to tell the agg how many buckets to expect and use that as the basis of calculations.  This would be somewhat similar to the `gap_policy`, which tells pipelines how to handle missing buckets.  Instead, we're just telling them how many buckets to expect.

Sorry for the unsatisfying answer.  :(

**Comment by LucidObscurity** [comment]
Thanks @polyfractal. I did really like your `"parent_bounds"` suggestion since it's simple and I can't imagine why anyone would want to extend past the parent. 

I wound up rolling a little .net core web api to get around this and some other issues (like grabbing only business hours or excluding weekends without corrupting any windowed statistics) which sits in-between the kibana vega visualization and elasticsearch. This way I can add and remove buckets as needed and I just rolled my own stats and moving averages. I was hoping to be able to scrap this little middleman at some point. 

When I get some free time I'll put in another feature request I think would be handy...a way to filter date histo buckets like I can use a bool script filter to filter docs, but for the buckets themselves (since I want the empty buckets, but not all of them). For instance `doc['@timestamp'].date.getDayOfWeek() < 6`. Anyway, I'll save the details for the next feature request.

I'll keep an eye on the release notes with my fingers crossed.

Thanks again.
